### PR TITLE
Use a more exact base class for considerGuardedDevirtualization

### PR DESCRIPTION
**Pull request by EgorBo**
An attempt to address https://github.com/dotnet/runtime/issues/108471#issuecomment-2389727329

### CODE REVIEW

**Comment by EgorBo** [comment]
@MihuBot

**Comment by MichalStrehovsky** [comment]
I ran a rt-sz run on this: https://github.com/MichalStrehovsky/rt-sz/issues/74#issuecomment-2390794436

This is a size regression but for good reasons. You can grab the before/after binaries from here: https://github.com/MichalStrehovsky/rt-sz/actions/runs/11158038241

I fed the webapiaot to Ghidra and then exported the analysis to BinDiff format and ran BinDiff on it to spot check the diffs. They are in the expected two categories - no longer devirtualizing things that were apparently impossible, and newly devirtualizing things that weren't.

The newly devirtualized cases look interesting. For example, I'm looking at a devirtualized `IDisposable.Dispose`. Previously if RyuJIT was asking who implements `IDisposable.Dispose`, it would get tons of answers to the point it's not useful. However if we restrict this to some known type, this seems to work better.

If you want to have a look, I'm putting the binexport files here:

[baseline.zip](https://github.com/user-attachments/files/17242156/baseline.zip)
[compare.zip](https://github.com/user-attachments/files/17242160/compare.zip)

You shouldn't need Ghidra for this, just download the latest BinDiff and these two binexport files and create a new diff from them.

File: src/coreclr/jit/importercalls.cpp [hunk 1]
```diff
@@ -8175,7 +8175,13 @@ void Compiler::impDevirtualizeCall(GenTreeCall*            call,
             return;
         }
 
-        considerGuardedDevirtualization(call, ilOffset, isInterface, baseMethod, baseClass, pContextHandle);
+        CORINFO_CLASS_HANDLE moreExactBaseClass = baseClass;
+        if (objClass != NO_CLASS_HANDLE)
```
Comment (AndyAyersMS) [inline src/coreclr/jit/importercalls.cpp:8179]: Don't we already know `objClass` is not null here?

**Comment by AndyAyersMS** [comment]
PGO stress failures look like #108360

File: src/coreclr/jit/importercalls.cpp (hunk 1 above)
Comment (EgorBo) [reply src/coreclr/jit/importercalls.cpp]: > Don't we already know `objClass` is not null here?

Yep, you're correct, removed the condition

**Review by AndyAyersMS — APPROVED** [review]

**Comment by JulieLeeMSFT** [comment]
Do we need to backport this?

**Comment by AndyAyersMS** [comment]
> Do we need to backport this?

It is a missed optimization, so I would guess doesn't meet the bar.

**Pull request merged**